x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationDenialMessages.java

```diff
@@ -90,22 +93,47 @@ public String actionDenied(
 
             if (ClusterPrivilegeResolver.isClusterAction(action)) {
                 final Collection<String> privileges = findClusterPrivilegesThatGrant(authentication, action, request);
-                if (privileges != null && privileges.size() > 0) {
+                if (privileges != null && false == privileges.isEmpty()) {
                     message = message
                         + ", this action is granted by the cluster privileges ["
                         + collectionToCommaDelimitedString(privileges)
                         + "]";
                 }
             } else if (isIndexAction(action)) {
-                final Collection<String> privileges = findIndexPrivilegesThatGrant(action);
-                if (privileges != null && privileges.size() > 0) {
+                // this includes `all`
+                final Collection<String> privileges = findIndexPrivilegesThatGrant(
+                    action,
+                    p -> p.getSelectorPredicate().test(IndexComponentSelector.DATA)
+                );
+                // this is an invariant since `all` is included in the above so the only way
+                // we can get an empty result here is a bogus action, which will never be covered by a failures privilege
+                assert false == privileges.isEmpty()
+                    || findIndexPrivilegesThatGrant(
+                        action,
+                        p -> p.getSelectorPredicate().test(IndexComponentSelector.FAILURES)
+                            && false == p.getSelectorPredicate().test(IndexComponentSelector.DATA)
+                    ).isEmpty()
+                    : "action [" + action + "] is not covered by any regular index privilege, only by failures-selector privileges";
+
+                if (false == privileges.isEmpty()) {
                     message = message
                         + ", this action is granted by the index privileges ["
                         + collectionToCommaDelimitedString(privileges)
                         + "]";
+
+                    final Collection<String> privilegesForFailuresOnly = findIndexPrivilegesThatGrant(
+                        action,
+                        p -> p.getSelectorPredicate().test(IndexComponentSelector.FAILURES)
+                            && false == p.getSelectorPredicate().test(IndexComponentSelector.DATA)
+                    );
+                    if (false == privilegesForFailuresOnly.isEmpty() && hasIndicesWithFailuresSelector(request)) {
+                        message = message
+                            + " for data access, or by ["
+                            + collectionToCommaDelimitedString(privilegesForFailuresOnly)
+                            + "] for access with the [failures] selector";
```

Comment:
optional: Should we go with `::failures` instead?
```suggestion
                            + "] for access with the [::failures] selector";
```